strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/nuget/NugetArtifactController.java

```diff
@@ -661,56 +666,54 @@ private void writePackagePart(String boundaryString,
+        InputStream inputForNuspec = new ByteArrayInputStream(byteArray);
+        InputStream inputForNupkg = new ByteArrayInputStream(byteArray);
 
-            try (InputStream bis = new BufferedInputStream(Files.newInputStream(nuspecFile)))
-            {
-                nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, bis);
-            }
+        nuspecFile = extractNuspecFile(inputForNuspec);
+        if (nuspecFile == null) {
+            return null;
+        }
+        path = String.format("%s/%s/%s.%s.nupkg",
+                nuspecFile.getId(),
+                nuspecFile.getVersion(),
+                nuspecFile.getId(),
+                nuspecFile.getVersion());
+        nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, inputForNupkg);
+        Path nuspecFileTmp = Files.createTempFile(nuspecFile.getId(), "nuspec");
+        try (OutputStream outputStream = Files.newOutputStream(nuspecFileTmp)) {
+            nuspecFile.saveTo(outputStream);
+        }
+        path = String.format("%s/%s/%s.nuspec", nuspecFile.getId(), nuspecFile.getVersion(), nuspecFile.getId());
 
-            Path hashFile = Files.createTempFile(String.format("%s.%s", nupkgFile.getId(), nupkgFile.getVersion()),
-                                                "nupkg.sha512");
+        nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, Files.newInputStream(nuspecFileTmp));
 
-            try (OutputStream bos = new BufferedOutputStream(Files.newOutputStream(hashFile)))
-            {
-                nupkgFile.getHash().saveTo(bos);
-            }
-
-            path = String.format("%s/%s/%s.%s.nupkg.sha512",
-                                 nupkgFile.getId(),
-                                 nupkgFile.getVersion(),
-                                 nupkgFile.getId(),
-                                 nupkgFile.getVersion());
+        return new URI("");
+    }
 
-            try (InputStream bis = new BufferedInputStream(Files.newInputStream(hashFile)))
-            {
-                nugetArtifactManagementService.validateAndStore(storageId, repositoryId, path, bis);
+    private NuspecFile extractNuspecFile(InputStream is) {
+        NuspecFile nuspecFile;
+        ZipInputStream zipInputStream;
+
+        try {
+            zipInputStream = new ZipInputStream(is);
+            ZipEntry zipEntry;
+            do {
+                zipEntry = zipInputStream.getNextEntry();
+            } while (zipEntry != null && !zipEntry.getName().endsWith(".nuspec"));
+            if (zipEntry != null) {
+                nuspecFile = NuspecFile.parseZipStream(zipInputStream);
+                return nuspecFile;
             }
+        } catch (IOException | org.carlspring.strongbox.nuget.file.NugetFormatException e) {
+            logger.error(String.format("Failed to parse Nupkg"), e);
         }
 
-        return new URI("");
+        return null;
```

Comment:
why do we need to return `null` here? maybe it's better to just to add `throws IOException` to the method declaration?